amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpAsyncFlow.scala

```diff
@@ -0,0 +1,332 @@
+import akka.stream.alpakka.amqp.{AmqpWriteSettings, WriteMessage, WriteResult}
+import akka.stream.stage._
+import com.rabbitmq.client.ConfirmCallback
+
+import scala.collection.immutable.TreeMap
+import scala.collection.mutable
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{Future, Promise}
+
+/**
+ * Internal API.
+ */
+@InternalApi private final case class AwaitingMessage[T](
+    tag: DeliveryTag,
+    passThrough: T,
+    var ready: Boolean = false
+)
+
+/**
+ * Internal API.
+ */
+@InternalApi private object AmqpAsyncFlowStageLogic {
+  type DeliveryTag = Long
+}
+
+/**
+ * Internal API.
+ *
+ * Base stage for AMQP flows with asynchronous confirmations.
+ */
+@InternalApi private sealed abstract class AmqpAsyncFlowStageLogic[T](
+    override val settings: AmqpWriteSettings,
+    bufferSize: Int,
+    confirmationTimeout: FiniteDuration,
+    promise: Promise[Done],
+    shape: FlowShape[WriteMessage[T], WriteResult[T]]
+) extends TimerGraphStageLogic(shape)
+    with AmqpConnectorLogic
+    with StageLogging {
+
+  private def in = shape.in
+  private def out = shape.out
+
+  private val exchange = settings.exchange.getOrElse("")
+  private val routingKey = settings.routingKey.getOrElse("")
+  private val exitQueue = mutable.Queue.empty[WriteResult[T]]
+  private var upstreamException: Option[Throwable] = None
+
+  override def whenConnected(): Unit = {
+    channel.confirmSelect()
+    channel.addConfirmListener(asAsyncCallback(onConfirmation), asAsyncCallback(onRejection))
+  }
+
+  private def asAsyncCallback(confirmCallback: (DeliveryTag, Boolean) => Unit): ConfirmCallback = {
+    val callback = getAsyncCallback[(DeliveryTag, Boolean)] {
+      case (tag: DeliveryTag, multiple: Boolean) => confirmCallback(tag, multiple)
+    }
+    new ConfirmCallback { // cant use function literal because it doesn't work with 2.11
+      override def handle(tag: DeliveryTag, multiple: Boolean): Unit = callback.invoke((tag, multiple))
+    }
+  }
+
+  private def onConfirmation(tag: DeliveryTag, multiple: Boolean): Unit = {
+    log.debug("Received confirmation for deliveryTag {} (multiple={}).", tag, multiple)
+
+    val dequeued: Iterable[AwaitingMessage[T]] = dequeueAwaitingMessages(tag, multiple)
+
+    dequeued.foreach(m => cancelTimer(m.tag))
+
+    pushOrEnqueueResults(
+      dequeued.map(m => WriteResult.confirmed(m.passThrough))
+    )
+  }
+
+  private def onRejection(tag: DeliveryTag, multiple: Boolean): Unit = {
+    log.debug("Received rejection for deliveryTag {} (multiple={}).", tag, multiple)
+
+    val dequeued: Iterable[AwaitingMessage[T]] = dequeueAwaitingMessages(tag, multiple)
+
+    dequeued.foreach(m => cancelTimer(m.tag))
+
+    pushOrEnqueueResults(
+      dequeued.map(m => WriteResult.rejected(m.passThrough))
+    )
+  }
+
+  private def pushOrEnqueueResults(results: Iterable[WriteResult[T]]): Unit = {
+    results.foreach(result => {
```

Comment:
I would prefer `results.foreach { result => }` syntax than `results.foreach(result => {})`, less parentheses, easier to read.